### PR TITLE
copr: Add collation utils

### LINKED ISSUE #6574: Add collation interfaces to Coprocessor
## Feature Request

Add collation interfaces to Coprocessor following https://github.com/pingcap/tidb/pull/14574

### PR DESCRIPTION
**Pull request by breeswish**
###  What have you changed?

This PR adds basic collation utils according to https://github.com/bb7133/proto-collations/blob/master/collations_test.go. Resolves #6574 

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

### CODE REVIEW

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+    [0x00D9, 0x00F9, 0x0055],
+    [0x00DA, 0x00FA, 0x0055],
+    [0x00DB, 0x00FB, 0x0055],
+    [0x00DC, 0x00FC, 0x0055],
+    [0x00DD, 0x00FD, 0x0059],
+    [0x00DE, 0x00FE, 0x00DE],
+    [0x0178, 0x00FF, 0x0059],
+];
+
+const GENERAL_CI_PLANE_01: [[u16; 3]; 256] = [
```
Comment (wjhuang2016) [inline components/tidb_query/src/expr_util/collation/utf8mb4.rs]: In MySQL, there are many planes general_ci used, please add it.

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+    [0x01F8, 0x01F9, 0x004E],
+    [0x01FA, 0x01FB, 0x0041],
+    [0x01FA, 0x01FB, 0x0041],
+    [0x01FC, 0x01FD, 0x00C6],
+    [0x01FC, 0x01FD, 0x00C6],
+    [0x01FE, 0x01FF, 0x00D8],
+    [0x01FE, 0x01FF, 0x00D8],
+];
+
+pub struct CollateUtf8Mb4GeneralCi;
```
Comment (wjhuang2016) [inline components/tidb_query/src/expr_util/collation/utf8mb4.rs]: How about `Utf8Mb4GeneralCiCollator`? Also change the trait `collation` to `collator`.

File: components/tidb_query/src/expr_util/collation/utf8mb4.rs (hunk 1 above)
Comment (breeswish) [reply components/tidb_query/src/expr_util/collation/utf8mb4.rs]: TiKV will follow TiDB's implementation. Meanwhile TiDB has not implemented, this is a sample interface in order to not blocking other issues.

Comment (wjhuang2016) [reply components/tidb_query/src/expr_util/collation/utf8mb4.rs]: ok

File: components/tidb_query/src/expr_util/collation/mod.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+mod utf8mb4;
```
Comment (wjhuang2016) [inline components/tidb_query/src/expr_util/collation/mod.rs]: why `utf8mb4` ?

Comment (breeswish) [reply components/tidb_query/src/expr_util/collation/mod.rs]: What's your suggested name?

Comment (wjhuang2016) [reply components/tidb_query/src/expr_util/collation/mod.rs]: `collate`?

Comment (breeswish) [reply components/tidb_query/src/expr_util/collation/mod.rs]: it will only contain utf8/utf8mb4 collations, in case of any future charsets.

Comment (wjhuang2016) [reply components/tidb_query/src/expr_util/collation/mod.rs]: That is strange, why not contain 'bin'?

Comment (breeswish) [reply components/tidb_query/src/expr_util/collation/mod.rs]: because bin is placed in `mod.rs`.

**Comment by TennyZhuang** [comment]
@sre-bot /run-all-tests

File: components/codec/src/number.rs [hunk 4]
```diff
@@ -917,6 +917,15 @@ pub trait NumberEncoder: BufferWriter {
         write!(self, v, 2, encode_u16_le)
     }
 
+    /// Writes an unsigned 16 bit integer `v` in big endian,
+    /// which is memory-comparable.
+    ///
+    /// This function is an alias to `write_u16`.
+    #[inline]
+    fn write_u16_be(&mut self, v: u16) -> Result<()> {
```
Comment (wjhuang2016) [inline components/codec/src/number.rs]: Why adding the number logic?

Comment (breeswish) [reply components/codec/src/number.rs]: since write_u16_be is missing and it is used when generating sort key for general_ci.

**Comment by breeswish** [comment]
Close in favour of #6566

**Pull request closed without merging**